File: app/src/main/java/social/entourage/android/view/HtmlTextView.java
package social.entourage.android.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.text.method.LinkMovementMethod;
import android.text.method.MovementMethod;
import android.util.AttributeSet;
import android.widget.TextView;

import social.entourage.android.R;
import social.entourage.android.tools.Utils;

public class HtmlTextView extends androidx.appcompat.widget.AppCompatTextView {

    private String htmlString;

    public HtmlTextView(Context context) {
        super(context);
    }

    public HtmlTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        buildAttrs(context, attrs);
    }

    public HtmlTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        buildAttrs(context, attrs);
    }

    private void buildAttrs(Context context, AttributeSet attrs) {
        TypedArray typedArray = context.getTheme().obtainStyledAttributes(attrs, R.styleable.HtmlTextView, 0, 0);
        try {
            htmlString = typedArray.getString(R.styleable.HtmlTextView_htmlText);
            if (htmlString != null) {
                setText(Utils.fromHtml(htmlString), TextView.BufferType.SPANNABLE);
                setMovementMethod(LinkMovementMethod.getInstance());
            }
        } finally {
            typedArray.recycle();
        }
    }

    public void setHtmlString(int resourceID) {
        setHtmlString(getResources().getString(resourceID));
    }

    public void setHtmlString(String htmlString) {
        setHtmlString(htmlString, LinkMovementMethod.getInstance());
    }

    public void setHtmlString(String htmlString, MovementMethod movementMethod) {
        if (htmlString == null) htmlString = "";
        setText(Utils.fromHtml(htmlString), BufferType.SPANNABLE);
        this.htmlString = htmlString;
        setMovementMethod(movementMethod);
    }
}

File: app/src/main/java/social/entourage/android/view/HtmlTextView.kt
package social.entourage.android.view

import android.content.Context
import android.text.method.LinkMovementMethod
import android.text.method.MovementMethod
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import social.entourage.android.R
import social.entourage.android.tools.Utils

class HtmlTextView : AppCompatTextView {
    private var htmlString: String? = null

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        buildAttrs(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        buildAttrs(context, attrs)
    }

    private fun buildAttrs(context: Context, attrs: AttributeSet?) {
        val typedArray = context.theme.obtainStyledAttributes(attrs, R.styleable.HtmlTextView, 0, 0)
        try {
            typedArray.getString(R.styleable.HtmlTextView_htmlText)?.let {
                htmlString = it
                setText(Utils.fromHtml(it), BufferType.SPANNABLE)
                movementMethod = LinkMovementMethod.getInstance()
            } ?: run {
                htmlString = null
            }
        } finally {
            typedArray.recycle()
        }
    }

    fun setHtmlString(resourceID: Int) {
        setHtmlString(resources.getString(resourceID))
    }

    fun setHtmlString(htmlString: String?) {
        setHtmlString(htmlString, LinkMovementMethod.getInstance())
    }

    fun setHtmlString(newHtmlString: String?, movementMethod: MovementMethod?) {
        this.htmlString = newHtmlString ?: ""
        setText(Utils.fromHtml(htmlString ?: ""), BufferType.SPANNABLE)
        this.movementMethod = movementMethod
    }
}